# Practice 1

In this exercise, you will warm up / refresh your Python skills by analyzing some real data inside of a Jupyter notebook environment. You should write whatever **Python** code you like to solve these problems. You can search on the web for documentation and techniques to help you solve the problem. See especially the official [Python tutorial](https://docs.python.org/3/tutorial/) and [documentation](https://docs.python.org/3/).

For help with Jupyter notebooks press `h` when not in edit mode for a cell or see the [documentation](https://jupyter-notebook.readthedocs.io/en/stable/).

You can either print answers directly from your code or write them in the markdown cells below your code. Either way, make sure that your answers are visible and can be easily read in the notebook you submit.

## Part 1: Text
The `text.txt` file contains the opening sentences of Charles' Dickens novel *A Tale of Two Cities*. Run the code below. It will read the file, store it in a string variable `text`, and print the first 500 characters.

In [26]:
f = open('text.txt')
text = f.read()
f.close()

print(text[:500], '...')

it was the best of times, it was the worst of times, it was the age of wisdom, it was the age of foolishness, it was the epoch of belief, it was the epoch of incredulity, it was the season of light, it was the season of darkness, it was the spring of hope, it was the winter of despair, we had everything before us, we had nothing before us, we were all going direct to heaven, we were all going direct the other way- in short, the period was so far like the present period, that some of its noisiest ...


### Question 1
Print the first and the last sentence of the text. You may assume sentences only end with a period (`'.'`).
You may find the Python `split` method for strings to be useful: https://docs.python.org/3/library/stdtypes.html?highlight=split#str.split `sep='.'`.

In [32]:
# Code for question 1

text2 = text.strip()
sentences = text2.split('.')
print(sentences[0])
print(sentences[-2])

it was the best of times, it was the worst of times, it was the age of wisdom, it was the age of foolishness, it was the epoch of belief, it was the epoch of incredulity, it was the season of light, it was the season of darkness, it was the spring of hope, it was the winter of despair, we had everything before us, we had nothing before us, we were all going direct to heaven, we were all going direct the other way- in short, the period was so far like the present period, that some of its noisiest authorities insisted on its being received, for good or for evil, in the superlative degree of comparison only
 thus did the year one thousand seven hundred and seventy-five conduct their greatnesses, and myriads of small creatures-the creatures of this chronicle among the rest-along the roads that lay before them


### Answer 1

First Sentence: it was the best of times, it was the worst of times, it was the age of wisdom, it was the age of foolishness, it was the epoch of belief, it was the epoch of incredulity, it was the season of light, it was the season of darkness, it was the spring of hope, it was the winter of despair, we had everything before us, we had nothing before us, we were all going direct to heaven, we were all going direct the other way- in short, the period was so far like the present period, that some of its noisiest authorities insisted on its being received, for good or for evil, in the superlative degree of comparison only

Second Sentence: thus did the year one thousand seven hundred and seventy-five conduct their greatnesses, and myriads of small creatures-the creatures of this chronicle among the rest-along the roads that lay before them

### Question 2
How many total vowels (`a`, `e`, `i`, `o`, `u`) appear in the text?

In [33]:
# Code for question 2

print(text.count('a') + text.count('e') + text.count('i') + text.count('o') + text.count('u'))

1735


### Answer 2
1735 vowels appear in the text.

### Question 3
A trigram is a string of exactly 3 characters (or a string with length exactly 3), including blank spaces. For example, the first five trigrams in the text are:
1. `'it '`
2. `'t w'`
3. `' wa'`
4. `'was'`
5. `'as '`

What is the most common trigram in the text?   

In [41]:
# Code for question 3

trigrams = {}

for i in range(len(text)):
    
    if text[i:i+3] not in trigrams:
        trigrams[text[i:i+3]] = 0
    trigrams[text[i:i+3]] += 1
    
m = max(trigrams.values())

for k in trigrams.keys():
    if trigrams[k] == m:
        print(k)

 th


### Answer 3
The most common trigram in the text is " th"

## Part 2: Rides

The [boston.csv](boston.csv) file contains data on weekday Uber rides in the Boston, Massachusetts metropolitan areas from the [Uber Movement](https://movement.uber.com) project. The `sourceid` and `dstid` columns contain codes corresponding to start and end locations of each ride. The `hod` column contains codes corresponding to the hour of the day the ride took place. The `ride time` column contains the length of the ride, in minutes.

The code below will open the file and read the data as a list of rows, with each row represented as a string. It then prints the first four rows. Note that the first (index 0) row contains the column headers. 

In [42]:
f = open('boston.csv')
data = f.readlines()
f.close()

for row in data[:4]:
    print(row)

sourceid,dstid,hod,ride time

584,33,7,11.866000000000001

1013,1116,13,17.799333333333333

884,1190,22,19.348833333333335



### Question 4
How many rides are listed in the the file?

In [48]:
# Code for question 4

print(len(data))

200001


### Answer 4
There are 200000 rides listed in the file.

### Question 5
What is the maximum length of a ride?

In [58]:
# Code for question 5

max = 0

for row in data[1:]:
    row2 = row.split(',')
    time = float(row2[3][0:-2])
    if time > max:
        max = time

print(max)

1471.986166666666


### Answer 5
The maximum length of a ride is 1471.99 minutes.

### Question 6
What is the average length of a ride?

In [60]:
# Code for question 6

times = []

for row in data[1:]:
    row2 = row.split(',')
    time = float(row2[3][0:-2])
    times.append(time)

print(sum(times)/len(times))

16.49678068750019


### Answer 6
The average length of a ride is 16.5 minutes.

### Question 7
What percentage of rides are under 10 minutes?

In [61]:
# Code for question 7

u = 0
for t in times:
     if t < 10:
            u += 1
print(u/len(times))

0.22829


### Answer 7
22.83% of rides are under 10 minutes.

### Question 8
For each HOD (hour of day), count how many rides take place during that hour.

In [69]:
# Code for question 8

hours = {}

for row in data[1:]:
    row2 = row.split(',')
    hour = float(row2[2])
    if hour not in hours:
        hours[hour] = 0
    hours[hour] += 1

for key in sorted(hours.keys()):
    print("%s : %s" % (key, hours[key]))

0.0 : 7600
1.0 : 6568
2.0 : 5555
3.0 : 4296
4.0 : 4371
5.0 : 5343
6.0 : 7424
7.0 : 9173
8.0 : 9510
9.0 : 9259
10.0 : 9053
11.0 : 8755
12.0 : 8885
13.0 : 8813
14.0 : 9688
15.0 : 9748
16.0 : 9930
17.0 : 9952
18.0 : 8365
19.0 : 8926
20.0 : 9771
21.0 : 9894
22.0 : 9863
23.0 : 9258


### Answer 8
0.0 : 7600  
1.0 : 6568  
2.0 : 5555  
3.0 : 4296  
4.0 : 4371  
5.0 : 5343  
6.0 : 7424  
7.0 : 9173  
8.0 : 9510  
9.0 : 9259  
10.0 : 9053  
11.0 : 8755  
12.0 : 8885  
13.0 : 8813  
14.0 : 9688  
15.0 : 9748  
16.0 : 9930  
17.0 : 9952  
18.0 : 8365  
19.0 : 8926  
20.0 : 9771  
21.0 : 9894  
22.0 : 9863  
23.0 : 9258  

### Question 9
What are the three most common **start** locations (`sourceid`) for rides in the dataset? You may find the built-in Python `sorted` function helpful; it's usage is described in https://docs.python.org/3/howto/sorting.html.

In [80]:
# Code for question 9

starts = {}

for row in data[1:]:
    row2 = row.split(',')
    start = float(row2[0])
    if start not in starts:
        starts[start] = 0
    starts[start] += 1

print(sorted(starts.items(), key = lambda item: item[1], reverse = True))

[(885.0, 1181), (498.0, 1067), (435.0, 1057), (1116.0, 1049), (1048.0, 1035), (884.0, 1031), (540.0, 980), (1038.0, 980), (856.0, 978), (1110.0, 971), (883.0, 956), (297.0, 952), (333.0, 893), (878.0, 893), (886.0, 883), (497.0, 878), (207.0, 875), (1193.0, 865), (879.0, 864), (53.0, 864), (487.0, 853), (866.0, 843), (1164.0, 828), (47.0, 816), (154.0, 799), (27.0, 790), (1179.0, 780), (861.0, 778), (21.0, 767), (1043.0, 751), (51.0, 747), (496.0, 746), (1040.0, 739), (290.0, 721), (882.0, 710), (1226.0, 707), (68.0, 706), (1140.0, 700), (203.0, 700), (1189.0, 675), (1127.0, 672), (302.0, 667), (1227.0, 663), (917.0, 660), (872.0, 657), (880.0, 655), (504.0, 655), (413.0, 654), (881.0, 650), (280.0, 647), (167.0, 647), (294.0, 635), (277.0, 632), (495.0, 628), (1002.0, 622), (94.0, 621), (286.0, 587), (448.0, 580), (760.0, 569), (1024.0, 568), (857.0, 564), (151.0, 563), (889.0, 558), (862.0, 557), (362.0, 556), (1214.0, 555), (210.0, 554), (757.0, 553), (503.0, 553), (888.0, 553), (40

### Answer 9
The three most common start locations are 885, 498, and 435.

## Submitting

You should make sure any code that you write to answer the questions is included in this notebook. Save your work. Double check that your entire notebook runs correctly and generates the expected output. To do so, you can simply select Kernel -> Restart and Run All. When you finish, submit your assignment at http://gradescope.com/.